src/ResourceManager/Common/Commands.ResourceManager.Common/Location/LocationCompleter.cs

```diff
@@ -0,0 +1,200 @@
+                                context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                                AzureSession.Instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                                AzureSession.Instance.ClientFactory.GetCustomHandlers());
+                            client.SubscriptionId = context.Subscription.Id;
+                            var allProviders = client.Providers.ListAsync();
+                            if (allProviders.Wait(TimeSpan.FromSeconds(5)))
+                                _resourceTypeLocationDictionary[contextHash] = CreateLocationDictionary(allProviders.Result.ToList());
+                            else
+                            {
+                                _resourceTypeLocationDictionary[contextHash] = new Dictionary<string, ICollection<string>>();
+#if DEBUG
+                                throw new Exception(Resources.TimeOutForProviderList);
+#endif
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            _resourceTypeLocationDictionary[contextHash] = new Dictionary<string, ICollection<string>>();
+#if DEBUG
+                            throw ex;
+#endif
+                        }
+                    }
+
+                    return _resourceTypeLocationDictionary[contextHash];
+                }
+            }
+        }
+
+        /// <summary>
+        /// Pass in a list of ResourceTypes and this class will provide a list of locations that are common to all ResourceTypes given. This will then be available to the user to tab through.
+        /// Example: [Parameter(ParameterSetName = ListByNameInTenantParameterSet, ValueFromPipelineByPropertyName = true, Mandatory = false), LocationCompleter(new string[] { "Microsoft.Batch/operationss" })]
+        /// </summary>
+        /// <param name="resourceTypes"></param>
+        public LocationCompleterAttribute(string[] resourceTypes) : base(CreateScriptBlock(resourceTypes))
+        {
+        }
+
+        public static string[] FindLocations(string[] resourceTypes)
+        {
+            return FindLocations(resourceTypes, ResourceTypeLocationDictionary);
+        }
+
+        /// <summary>
+        /// Finds all valid Locations for 
+        /// </summary>
+        /// <param name="resourceTypes"></param>
+        /// <param name="resourceTypeLocationDictionary"></param>
+        /// <returns></returns>
+        public static string[] FindLocations(string[] resourceTypes, IDictionary<string, ICollection<string>> resourceTypeLocationDictionary)
+        {
+            if (resourceTypeLocationDictionary == null)
+            {
+                resourceTypeLocationDictionary = new Dictionary<string, ICollection<string>>();
```

Comment:
Since multiple threads may use this dictionary, please use ConcurrentDictionary by default.

Reply:
ALso, always declare dictionaries with a KweCompartor (in this case StringComparsion.OrdinalIgnoreCase